src/block-ref.ts

```diff
@@ -0,0 +1,60 @@
+import {
+  createAsyncMiddleware,
+  JsonRpcMiddleware,
+  PendingJsonRpcResponse,
+} from 'json-rpc-engine';
+import clone from 'clone';
+import pify from 'pify';
+import { blockTagParamIndex } from './cache-utils';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires,@typescript-eslint/no-require-imports
+const BlockTracker = require('eth-block-tracker');
+
+interface BlockRefMiddlewareOptions{
+  blockTracker?: typeof BlockTracker;
+  provider?: Record<string, unknown>;
```

Comment:
We should create an interface for `provider` with the properties we expect. I think it's just a single function, `sendAsync`, with a `JsonRpcRequest` as its first parameter and an error-first callback as the second. It could be genericized, e.g.:
```typescript
interface Provider {
  sendAsync<T, U>: (JsonRpcRequest<T>, (err: Error, result: JsonRpcResponse<U>) => void) => void;
}
```